File: CSCI5551-Robotic/Homework/HW3/hw3_provided_code/teleoperation.cpp
// This code contains the getch function, which will get a keypress for you. You'll need to write the rest of the teleoperation node yourself.
// Remember the typical ROS node setup: init the node, set up the publisher, run a loop (probably getting the key press every time you loop), then publish messages based off of that.

#include <stdio.h>
#include <unistd.h>
#include <termios.h>

char key(' ');

// For non-blocking keyboard inputs
int getch(void)
{
  int ch;
  struct termios oldt;
  struct termios newt;

  // Store old settings, and copy to new settings
  tcgetattr(STDIN_FILENO, &oldt);
  newt = oldt;

  // Make required changes and apply the settings
  newt.c_lflag &= ~(ICANON | ECHO);
  newt.c_iflag |= IGNBRK;
  newt.c_iflag &= ~(INLCR | ICRNL | IXON | IXOFF);
  newt.c_lflag &= ~(ICANON | ECHO | ECHOK | ECHOE | ECHONL | ISIG | IEXTEN);
  newt.c_cc[VMIN] = 1;
  newt.c_cc[VTIME] = 0;
  tcsetattr(fileno(stdin), TCSANOW, &newt);

  // Get the current character
  ch = getchar();

  // Reapply old settings
  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);

  return ch;
}

int main(int argc, char** argv)
{
 
  while(true){

    // Get the pressed key
    key = getch();
    printf("%c\n", key);
   
    if (key == '\x03')
    {
      break;
    }
  }

  return 0;
}

File: CSCI5551-Robotic/Homework/HW3/hw3_provided_code/teleoperation.py
#!/usr/bin/env python
#
# This code contains the getKey function, which will get a keypress for you. You'll need to write the rest of the teleoperation node yourself.
# Remember the typical ROS node setup: init the node, set up the publisher, run a loop (probably getting the key press every time you loop), then publish messages based off of that.


import sys, select, termios, tty


def getKey(key_timeout):
    tty.setraw(sys.stdin.fileno())
    rlist, _, _ = select.select([sys.stdin], [], [], key_timeout)
    if rlist:
        key = sys.stdin.read(1)
    else:
        key = ''
    termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)
    return key

if __name__=="__main__":
    settings = termios.tcgetattr(sys.stdin)
    
    while(1):
        key = getKey(0.1)
        # If keypress is not empty, print out the key.
        if key != '':
            print(key)

        # If keypress is Crtl+C, break loop and exit.
        if key == '\x03':
            break
        
    termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)
